src/bootstrap/native.rs

```diff
@@ -858,3 +858,62 @@ impl HashStamp {
         fs::write(&self.path, self.hash.as_deref().unwrap_or(b""))
     }
 }
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
+pub struct CrtBeginEnd {
+    pub target: TargetSelection,
+}
+
+impl Step for CrtBeginEnd {
+    type Output = PathBuf;
+
+    fn should_run(run: ShouldRun<'_>) -> ShouldRun<'_> {
+        run.path("src/llvm-project/compiler-rt/lib/crt")
+    }
+
+    fn make_run(run: RunConfig<'_>) {
+        run.builder.ensure(CrtBeginEnd { target: run.target });
+    }
+
+    /// Build crtbegin.o/crtend.o for musl target.
+    fn run(self, builder: &Builder<'_>) -> Self::Output {
+        let out_dir = builder.native_dir(self.target).join("crt");
+
+        if builder.config.dry_run {
+            return out_dir;
+        }
+
+        let done_stamp = out_dir.join("crt-finished-building");
+        if done_stamp.exists() {
+            return out_dir;
+        }
+
+        builder.info("Building crtbegin.o and crtend.o");
+        t!(fs::create_dir_all(&out_dir));
+
+        let mut cfg = cc::Build::new();
+
+        if let Some(ar) = builder.ar(self.target) {
+            cfg.archiver(ar);
+        }
+        cfg.compiler(builder.cc(self.target));
+        cfg.cargo_metadata(false)
+            .out_dir(&out_dir)
+            .target(&self.target.triple)
+            .host(&builder.config.build.triple)
+            .warnings(false)
+            .debug(false)
+            .opt_level(3)
+            .file(builder.src.join("src/llvm-project/compiler-rt/lib/crt/crtbegin.c"))
+            .file(builder.src.join("src/llvm-project/compiler-rt/lib/crt/crtend.c"))
+            .define("CRT_HAS_INITFINI_ARRAY", None)
+            .define("EH_USE_FRAME_REGISTRY", None)
+            .compile("crt");
+
+        t!(File::create(&done_stamp));
+        t!(fs::remove_file(out_dir.join("libcrt.a")));
```

Comment:
Does `cc` support using `cfg.flag("-c")` to only produce object files, but not a library or executable?
In that case specifying archiver or dealing with `.a` won't be necessary.

(Anyway, removing `libcrt.a` should be unnecessary in any case.)

Reply:
> Does cc support using cfg.flag("-c") to only produce object files, but not a library or executable?

The public api don't support this. https://github.com/alexcrichton/cc-rs/blob/1f50999ffb360db0f21cf166cc4cb580c8bc379c/src/lib.rs#L965